addons/l10n_in_edi_ewaybill/models/account_edi_format.py

```diff
@@ -0,0 +1,613 @@
+            error_message.append(_("Minimum one goods line is required to create E-waybill"))
+        if error_message:
+            error_message.insert(0, "For Ewaybill")
+        return error_message
+
+    def _post_invoice_edi(self, invoices):
+        if self.code != "in_ewaybill_1_03":
+            return super()._post_invoice_edi(invoices)
+        base = self._l10n_in_edi_ewaybill_base_irn_or_direct(invoices)
+        if base == "irn":
+            return self._l10n_in_edi_ewaybill_irn_post_invoice_edi(invoices)
+        else:
+            return self._l10n_in_edi_ewaybill_post_invoice_edi(invoices)
+
+    def _cancel_invoice_edi(self, invoices):
+        if self.code != "in_ewaybill_1_03":
+            return super()._cancel_invoice_edi(invoices)
+        response = {}
+        res = {}
+        ewaybill_response_json = invoices._get_l10n_in_edi_ewaybill_response_json()
+        cancel_json = {
+            "ewbNo": ewaybill_response_json.get("ewayBillNo") or ewaybill_response_json.get("EwbNo"),
+            "cancelRsnCode": int(invoices.l10n_in_edi_cancel_reason),
+            "CnlRem": invoices.l10n_in_edi_cancel_remarks,
+        }
+        response = self._l10n_in_edi_ewaybill_cancel(invoices.company_id, cancel_json)
+        if response.get("error"):
+            error = response["error"]
+            error_codes = [e.get("code") for e in error]
+            if "238" in error_codes:
+                # Invalid token eror then create new token and send generate request again.
+                # This happen when authenticate called from another odoo instance with same credentials (like. Demo/Test)
+                authenticate_response = self._l10n_in_edi_ewaybill_authenticate(invoices.company_id)
+                if not authenticate_response.get("error"):
+                    error = []
+                    response = self._l10n_in_edi_ewaybill_cancel(invoices.company_id, cancel_json)
+                    if response.get("error"):
+                        error = response["error"]
+                        error_codes = [e.get("code") for e in error]
+            if "312" in error_codes:
+                # E-waybill is alrady canceled
+                # this happens when timeout from the Government portal but IRN is generated
+                error_message = "<br/>".join(["[%s] %s" % (e.get("code"), html_escape(e.get("message"))) for e in error])
+                error = []
+                response = {"data": ""}
+                odoobot = self.env.ref("base.partner_root")
+                invoices.message_post(author_id=odoobot.id, body=markupsafe.Markup(_(
+                    "Somehow this Ewaybill is canceled in government portal before." \
+                    "<br/>Normally, this should not happen too often" \
+                    "<br/>Just verify by enter details into government website " \
+                    "<a href='https://ewaybillgst.gov.in/Others/EBPrintnew.aspx'>here<a>." \
+                    "<br/>Error: %s",
```

Comment:
This should not happen too often -> Please check this is not because of a bad configuration issue. 
You can verify by checking the details in the government website:

Reply:
same code for the error duplicated several times => shouldn't that be handled in a function to be factorized?